### PR TITLE
[String] fix truncating in WordBefore mode with length after last space

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | 
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/String/Tests/AbstractAsciiTestCase.php [hunk 1]
```diff
@@ -1536,12 +1536,24 @@ public static function provideTruncate(): array
             ['foobar foo aar', 'foobar foo aar', 12, '...', false],
             ['foobar...', 'foobar foo', 6, '...', TruncateMode::WordAfter],
             ['foobar...', 'foobar foo', 7, '...', TruncateMode::WordAfter],
+            ['foobar...', 'foobar foo a', 10, '...', TruncateMode::Char],
+            ['foobar...', 'foobar foo a', 10, '...', TruncateMode::WordBefore],
             ['foobar foo...', 'foobar foo a', 10, '...', TruncateMode::WordAfter],
+            ['foobar fo...', 'foobar foo aar', 12, '...', TruncateMode::Char],
             ['foobar foo aar', 'foobar foo aar', 12, '...', TruncateMode::WordAfter],
+            ['foobar...', 'foobar foo aar', 12, '...', TruncateMode::WordBefore],
```
Comment (xabbuh) [inline src/Symfony/Component/String/Tests/AbstractAsciiTestCase.php:1560]: this is the test case that fails without the bugfix

Comment (xabbuh) [reply src/Symfony/Component/String/Tests/AbstractAsciiTestCase.php:1560]: I added the other test cases to help understand the difference between the different modes applied on different input strings with different lengths

**Review by yceruto — APPROVED** [review]

File: src/Symfony/Component/String/Tests/AbstractAsciiTestCase.php [hunk 2]
```diff
@@ -1534,14 +1534,42 @@ public static function provideTruncate(): array
             ['foobar...', 'foobar foo', 7, '...', false],
             ['foobar foo...', 'foobar foo a', 10, '...', false],
             ['foobar foo aar', 'foobar foo aar', 12, '...', false],
+            ['foobar', 'foobar foo', 6, '', TruncateMode::Char],
+            ['foobar', 'foobar foo', 6, '', TruncateMode::WordAfter],
+            ['foobar', 'foobar foo', 6, '', TruncateMode::WordBefore],
+            ['foo...', 'foobar foo', 6, '...', TruncateMode::Char],
             ['foobar...', 'foobar foo', 6, '...', TruncateMode::WordAfter],
+            ['foobar...', 'foobar foo', 7, '', TruncateMode::WordBefore],
+            ['foobar ', 'foobar foo', 7, '', TruncateMode::Char],
```
Comment (xabbuh) [inline src/Symfony/Component/String/Tests/AbstractAsciiTestCase.php:1543]: As far as I can see this is the behaviour that we already had before the introduction of the truncate mode. Though I wonder if it really is intended to keep trailing spaces in this case.

**Comment by xabbuh** [comment]
there are some edge cases that are still not covered it seems

Status: Needs work

File: src/Symfony/Component/String/AbstractString.php [hunk 3]
```diff
@@ -620,14 +620,17 @@ public function truncate(int $length, string $ellipsis = '', bool|TruncateMode $
...
+        if (TruncateMode::WordAfter === $cut || !$cut) {
             if (null === $length = $this->indexOf([' ', "\r", "\n", "\t"], ($length ?: 1) - 1)) {
                 return clone $this;
             }
 
             $length += $ellipsisLength;
         }
 
+        if (TruncateMode::WordBefore) {
+        }
```
Comment (OskarStark) [inline src/Symfony/Component/String/AbstractString.php:631]: ?

Comment (xabbuh) [reply src/Symfony/Component/String/AbstractString.php:631]: not complete fix for https://github.com/symfony/symfony/pull/57778#issuecomment-2239294816 🙈

**Comment by xabbuh** [comment]
> there are some edge cases that are still not covered it seems

fixed

Status: Needs Review

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**